react/DeviceCoordinates.tsx

```diff
@@ -1,107 +1,113 @@
-import React, { useState, useEffect, useCallback } from 'react'
+import React, { useState, useEffect, useCallback, useMemo } from 'react'
 import { useAddressContext } from 'vtex.address-context/AddressContext'
 import { ButtonPlain, Spinner, Tooltip, IconLocation } from 'vtex.styleguide'
 import { FormattedMessage } from 'react-intl'
 import { useLazyQuery } from 'react-apollo'
 
 import REVERSE_GEOCODE_QUERY from './graphql/reverseGeocode.graphql'
 
-enum State {
+enum PermissionState {
   PROMPT,
+  PENDING,
   GRANTED,
-  LOADING,
   DENIED,
 }
 
 const DeviceCoordinates: StorefrontFunctionComponent = () => {
   const { setAddress } = useAddressContext()
-  const [state, setState] = useState<State>(State.PROMPT)
+  const [geolocationPermission, setGeolocationPermission] = useState<
+    PermissionState
+  >(PermissionState.PROMPT)
   const [executeReverseGeocode, { error, loading, data }] = useLazyQuery(
     REVERSE_GEOCODE_QUERY
   )
 
   useEffect(() => {
     if (data) {
       setAddress(data.reverseGeocode)
+    } else if (error) {
+      console.warn(error.message)
     }
-
-    if (error) {
-      console.warn(`error ${error.message}`)
-    }
-  }, [data, error, setAddress])
+  }, [data, error, loading, setAddress])
 
   const onGetCurrentPositionSuccess = useCallback(
     ({ coords }: Position) => {
+      setGeolocationPermission(PermissionState.GRANTED)
+
       executeReverseGeocode({
         variables: {
           lat: coords.latitude.toString(),
           lng: coords.longitude.toString(),
         },
       })
-
-      setState(State.GRANTED)
     },
     [executeReverseGeocode]
   )
 
   const onGetCurrentPositionError = useCallback((err: PositionError) => {
-    setState(State.DENIED)
+    setGeolocationPermission(PermissionState.DENIED)
     console.warn(`ERROR(${err.code}): ${err.message}`)
   }, [])
 
-  const onButtonClick = useCallback(() => {
-    setState(State.LOADING)
+  const requestGeolocation = useCallback(() => {
+    setGeolocationPermission(PermissionState.PENDING)
     navigator.geolocation.getCurrentPosition(
       onGetCurrentPositionSuccess,
       onGetCurrentPositionError,
       {
         enableHighAccuracy: true,
       }
     )
-  }, [onGetCurrentPositionSuccess, onGetCurrentPositionError])
+  }, [onGetCurrentPositionError, onGetCurrentPositionSuccess])
+
+  const handleButtonClick = () => {
+    requestGeolocation()
+  }
 
   useEffect(() => {
     navigator.permissions
       .query({ name: 'geolocation' })
-      .then((result: { state: string }) => {
+      .then((result: PermissionStatus) => {
         if (result.state === 'granted') {
```

Comment:
Why you didn't choose to use `PermissionState.GRANTED` here? Or another constant with those values.

Reply:
because `PermissionState` doesn't map to this value, as it is a value returned from the browser API. I could extract it to a constant, but I don't think it'd help much as the string is very descriptive

Reply:
Understood.